### PR TITLE
Multilevel Load balancing

### PR DESCRIPTION
**Pull request by rkotti**
### Contributing guide

https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md
### Changelog
- New trafficmanager-generic-arm template
  *
  *
  *
  *
### Description of the change

This is new traffic manager generic arm template segment that supports multilevel load balancing options from DSN load balancing(Traffic manager), L4 load balancing(Internet facing/internal load balancing) and L7 load balancing(Azure App Gateway).

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@rkotti__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (ram.kotti). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: .gitattributes [hunk 1]
```diff
@@ -0,0 +1,63 @@
+###############################################################################
```
Comment (gatneil) [inline .gitattributes]: Why is this file here? I doubt this was meant to be included...

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,1073 @@
+﻿{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "Region1-location-name": {
+      "defaultValue": "Central US",
+      "allowedValues": [
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: It's probably best to not have allowedValues for regions for 2 reasons:

1) different subscriptions will have different lists of regions they can use; listing all of them invites the user to accidentally choose a region they can't deploy to.
2) the list changes over time :)

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,1073 @@
...
+      }
+    },
+
+    "imageSKU": {
+      "type": "string",
+      "defaultValue": "2012-R2-Datacenter",
+      "allowedValues": [
+        "2008-R2-SP1",
+        "2012-Datacenter",
+        "2012-R2-Datacenter"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: No 2016? Not really required, but customers do tend to ask for it now that it's available

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,1073 @@
...
+    "App-prefix": {
+      "defaultValue": "ab",
+      "type": "string",
+      "minLength": 1,
+      "maxLength": 3,
+      "metadata": {
+        "description": "Application prefix name, should be < 3 characters"
+      }
+    },
+
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Please remove extra whitespace

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,1073 @@
...
+      "metadata": {
+        "description": "The Windows version for the VM"
+      }
+    },
+    "vmSize": {
+      "type": "string",
+      "minLength": 1,
+      "defaultValue": "Standard_D1_v2",
+      "allowedValues": [
+        "Basic_A0",
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Are we sure we want basic in this list? I thought basic didn't support load balancing: https://azure.microsoft.com/en-us/blog/basic-tier-virtual-machines-2/

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,1073 @@
...
+      "maxValue": 10,
+      "defaultValue": 4,
+      "metadata": {
+        "description": "Number of Application Gateway instances"
+      }
+    },
+    "appGW-pathMatch1": {
+      "type": "string",
+      "metadata": {
+        "description": "Path match string for Path Rule 1"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Is this description descriptive enough that people will know what to specify for it? It doesn't have a default; I don't know what I would put here...

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,1073 @@
...
+      }
+    },
+    "appGW-pathMatch2": {
+      "type": "string",
+      "metadata": {
+        "description": "Path match string for Path Rule 2"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: same

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,1073 @@
...
+      }
+    }
+  },
+  "variables": {
+    "region": [
+      "[parameters('Region1-location-name')]",
+      "[parameters('Region2-location-name')]"
+    ],
+    "apiVersion": "2015-06-15",
+    "CustomScriptScriptFileName": "CustomScript.ps1",
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Please see the best practice around staging artifacts': https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,1073 @@
...
+      "sku": {
+        "name": "Standard_LRS",
+        "tier": "Standard"
+      },
+      "kind": "Storage",
+      "location": "[variables('region')[copyIndex()]]",
+      "tags": {
+        "displayName": "storage-accounts"
+      },
+      "comments": "Traffic Manager Two Regions Demo.",
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: what is this? I don't think it's adding much extra description in this context

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,1073 @@
...
+      "dependsOn": [],
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Could we remove these empty lists and objects?

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,1073 @@
...
+            "publisher": "MicrosoftWindowsServer",
+            "offer": "WindowsServer",
+            "sku": "2012-R2-Datacenter",
+            "version": "latest"
+          },
+          "osDisk": {
+            "name": "[concat(variables('region-tm-test-vm'),copyIndex())]",
+            "createOption": "FromImage",
+            "vhd": {
+              "uri": "[concat('https', '://', variables('region-storage-account'),copyIndex(),'.blob.core.windows.net', concat('/vhds/',variables('region-tm-test-vm'),copyIndex(),'.vhd'))]"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Please see the best practice around 'uris compatable with all clouds': https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,1073 @@
...
+              "publicIPAddress": {
+                "id": "[resourceId('Microsoft.Network/publicIPAddresses',concat(variables('region1-web-vm-pip'), copyindex()))]"
+              }
+            }
+          }
+        ]
+      }
+    },
+
+
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Please remove extra spaces

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,1073 @@
...
+          "imageReference": {
+            "publisher": "[variables('imagePublisher')]",
+            "offer": "[variables('imageOffer')]",
+            "sku": "[parameters('imageSKU')]",
+            "version": "latest"
+          },
+          "osDisk": {
+            "name": "osdisk",
+            "vhd": {
+              "uri": "[concat('https://', variables('region-storage-account'),'0', '.blob.core.windows.net/vhds/', 'osdisk', copyindex(), '.vhd')]"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: best practice on uris for all clouds

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,1073 @@
...
+          "networkInterfaces": [
+            {
+              "id": "[resourceId('Microsoft.Network/networkInterfaces', concat(variables('region1-web-vm-nic'), copyindex()))]"
+            }
+          ]
+        },
+        "diagnosticsProfile": {
+          "bootDiagnostics": {
+            "enabled": true,
+            "storageUri": "[concat('https://', variables('region-storage-account'),'0', '.blob.core.windows.net')]"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: best practice on uris for all clouds

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,1073 @@
...
+          "imageReference": {
+            "publisher": "[variables('imagePublisher')]",
+            "offer": "[variables('imageOffer')]",
+            "sku": "[parameters('imageSKU')]",
+            "version": "latest"
+          },
+          "osDisk": {
+            "name": "osdisk",
+            "vhd": {
+              "uri": "[concat('https://', variables('region-storage-account'),'1', '.blob.core.windows.net/vhds/', 'osdisk', copyindex(), '.vhd')]"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: best practice on uris for all clouds

File: 301-multi-tier-loadbalancing/azuredeploy.json [hunk 16]
```diff
@@ -0,0 +1,1073 @@
...
+          "networkInterfaces": [
+            {
+              "id": "[resourceId('Microsoft.Network/networkInterfaces', concat(variables('region2-web-vm-nic'), copyindex()))]"
+            }
+          ]
+        },
+        "diagnosticsProfile": {
+          "bootDiagnostics": {
+            "enabled": true,
+            "storageUri": "[concat('https://', variables('region-storage-account'),'1', '.blob.core.windows.net')]"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: best practice on uris for all clouds

File: 301-multi-tier-loadbalancing/azuredeploy.parameters.json [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      "value": "North Europe"
+    },
+    "trafficManagerProfile_name": {
+      "value": "tmcptest"
+    },
+    "adminuser": {
+      "value": "admin123"
+    },
+    "adminuserPassword": {
+      "value": "TestUser@123"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.parameters.json]: Please use GEN-PASSWORD: https://github.com/Azure/azure-quickstart-templates/tree/master/1-CONTRIBUTION-GUIDE

File: 301-multi-tier-loadbalancing/azuredeploy.parameters.json [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    },
+
+    "App-prefix": {
+      "value": "cp"
+    },
+    "loadbalancer-dns-prefix": {
+      "value": "lb"
```
Comment (gatneil) [inline 301-multi-tier-loadbalancing/azuredeploy.parameters.json]: should probably use GEN-UNIQUE: https://github.com/Azure/azure-quickstart-templates/tree/master/1-CONTRIBUTION-GUIDE

**Comment by gatneil** [comment]
Please address the above comments :)

File: 301-multi-tier-loadbalancing/azuredeploy.parameters.json (hunk 17 above)
Comment (rkotti) [inline 301-multi-tier-loadbalancing/azuredeploy.parameters.json]: Done

File: 301-multi-tier-loadbalancing/azuredeploy.parameters.json (hunk 18 above)
Comment (rkotti) [inline 301-multi-tier-loadbalancing/azuredeploy.parameters.json]: Done

File: .gitattributes (hunk 1 above)
Comment (gatneil) [inline .gitattributes]: the .gitattributes file

Comment (rkotti) [inline .gitattributes]: I didn't find any file in this folder. Am I missing something?

Comment (gatneil) [inline .gitattributes]: How are you submitting this Pull Request? Is it through Visual Studio or some other IDE? It's possible this file is being generated by whichever IDE you're using; merging this in to the root of the repo will affect everyone, so I can't merge this file in...

Comment (singhkays) [inline .gitattributes]: @rkotti This file has been added at the root of the repo in your fork. I'd suggest removing it from there and then committing again to update the PR

File: 301-multi-tier-loadbalancing/azuredeploy.json (hunk 2 above)
Comment (singhkays) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: @rkotti The parameter naming style isn't consistent. Some of them follow camelCase and others follow some other type. Can you make this consistent?

File: .gitattributes (hunk 1 above)
Comment (rkotti) [inline .gitattributes]: Deleted! Thanks for pointing me exact location.

File: 301-multi-tier-loadbalancing/azuredeploy.json (hunk 2 above)
Comment (rkotti) [inline 301-multi-tier-loadbalancing/azuredeploy.json]: Parameters consistency has been addressed. 
No. Just took it random location.

**Comment by singhkays** [comment]
lgtm, merged
